src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -248,52 +292,174 @@ public static OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int b
                 destIndex += 3;
             }
-            else if (i2 != EncodingPad)
+            else if (t2 != EncodingPad)
             {
-                i2 = Unsafe.Add(ref decodingMap, i2);
+                int i2 = Unsafe.Add(ref decodingMap, (IntPtr)t2);
 
                 i2 <<= 6;
 
                 i0 |= i2;
 
                 if (i0 < 0)
                     goto InvalidExit;
-                Unsafe.Add(ref bufferBytes, destIndex) = (byte)(i0 >> 16);
-                Unsafe.Add(ref bufferBytes, destIndex + 1) = (byte)(i0 >> 8);
+
+                Unsafe.Add(ref bufferBytes, (IntPtr)destIndex) = (byte)(i0 >> 16);
+                Unsafe.Add(ref bufferBytes, (IntPtr)(destIndex + 1)) = (byte)(i0 >> 8);
                 destIndex += 2;
             }
             else
             {
                 if (i0 < 0)
                     goto InvalidExit;
-                Unsafe.Add(ref bufferBytes, destIndex) = (byte)(i0 >> 16);
+
+                Unsafe.Add(ref bufferBytes, (IntPtr)destIndex) = (byte)(i0 >> 16);
                 destIndex += 1;
             }
 
         DoneExit:
-            bytesWritten = destIndex;
+            bytesWritten = (int)destIndex;
             return OperationStatus.Done;
 
         InvalidExit:
-            bytesWritten = destIndex;
+            bytesWritten = (int)destIndex;
             return OperationStatus.InvalidData;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Avx2Decode(ref byte src, ref byte destBytes, int sourceLength, int destLength, ref uint sourceIndex, ref uint destIndex)
+        {
+            ref byte srcStart = ref src;
+            ref byte destStart = ref destBytes;
+            ref byte simdSrcEnd = ref Unsafe.Add(ref src, (IntPtr)((uint)sourceLength - 45 + 1));
+
+            // The JIT won't hoist these "constants", so help him
+            Vector256<sbyte> lutHi = s_avxDecodeLutHi;
+            Vector256<sbyte> lutLo = s_avxDecodeLutLo;
+            Vector256<sbyte> lutShift = s_avxDecodeLutShift;
+            Vector256<sbyte> mask2F = s_avxDecodeMask2F;
+            Vector256<sbyte> shuffleConstant0 = Vector256.Create(0x01400140).AsSByte();
+            Vector256<short> shuffleConstant1 = Vector256.Create(0x00011000).AsInt16();
+            Vector256<sbyte> shuffleVec = s_avxDecodeShuffleVec;
+            Vector256<int> permuteVec = s_avxDecodePermuteVec;
+
+            //while (remaining >= 45)
+            do
+            {
+                AssertRead<Vector256<sbyte>>(ref src, ref srcStart, sourceLength);
+                Vector256<sbyte> str = Unsafe.As<byte, Vector256<sbyte>>(ref src);
+
+                Vector256<sbyte> hiNibbles = Avx2.And(Avx2.ShiftRightLogical(str.AsInt32(), 4).AsSByte(), mask2F);
+                Vector256<sbyte> loNibbles = Avx2.And(str, mask2F);
+                Vector256<sbyte> hi = Avx2.Shuffle(lutHi, hiNibbles);
```

Comment:
note: 12.11 of the `Intel 64 and IA-32 Architectures Optimization Reference Manual` suggests replacing `Shuffle` with `Blend` where possible, as `Blend` can execute on an additional port and reduce pressure/improve performance.

> Assembly/Compiler Coding Rule 74. (M impact, M generality) Use Blend instructions in lieu of shuffle instruction in AVX whenever possible.

Reply:
Thx, I know this recommendation, but I couldn't came up with a blend-solution.

Here the shuffle is used to rearrange the lookup table (lut) `lutHi` according the (perfect) hash in `hiNibbles`.

Do you haven any suggestion on how to use blend on this case?

Reply:
I'd have to look into this more in-detail to determine if there is a good replacement.

Just wanted to callout the rule as something to be aware of for the time being.